### PR TITLE
Replace `ArrayList` usage in `ToolStripItemDataObject` with `List`

### PR DESCRIPTION
**Pull request by halgab**
Contributes to #8140

## Proposed changes

- Replace `ArrayList` usage in `ToolStripItemDataObject` with `List`
- Update relevant call sites to move off `IList` and `ICollection` interfaces

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/10626)

### CODE REVIEW

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ToolStripMenuItemDesigner.cs [hunk 1]
```diff
@@ -2655,7 +2655,7 @@ public override void OnDragDrop(Glyph g, DragEventArgs e)
                 if (ownerItem is not null && host is not null)
                 {
                     string transDesc;
-                    IList components = data.DragComponents;
+                    List<ToolStripItem> dragComponents = data.DragComponents;
                     int primaryIndex = -1;
                     bool copy = (e.Effect == DragDropEffects.Copy);
                     if (components.Count == 1)
```
Comment (elachlan) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ToolStripMenuItemDesigner.cs:2661]: Should be `dragComponents`?

**Comment by elachlan** [comment]
```
ToolStripItemBehavior.cs(656,119): error CS0103: (NETCORE_ENGINEERING_TELEMETRY=Build) The name 'components' does not exist in the current context
```

**Comment by halgab** [comment]
As mentioned [here](https://github.com/dotnet/winforms/pull/10577#discussion_r1446696061), this change might not be a no-brainer though...

**Comment by elachlan** [comment]
ToolStripMenuItemDesigner in the DemoConsole has had problems testing in the past see #9413

**Comment by elachlan** [comment]
@Olina-Zhang Is it possible for your team to test this? (I am not sure if it is, because demo console might not enable testing this)

I am concerned about the changes to `ToolStripItemDataObject` which is used "when a ToolStripItem is Drag-Dropped during DesignTime".

**Comment by halgab** [comment]
Abandoning this pr altogether is also a very acceptable outcome if we're not comfortable with it

**Comment by Olina-Zhang** [comment]
@elachlan @halgab We haven't started the test yet, should we run the test for this PR?

**Comment by halgab** [comment]
If you don't mind doing it, of course, please go ahead. Thanks a lot

**Comment by Olina-Zhang** [comment]
Tested this PR with some toolStripItems drag and drop scenarios in design time, no new issue found.

https://github.com/dotnet/winforms/assets/26474449/db3d753f-b656-456f-ab77-fa76424374ec

**Comment by elachlan** [comment]
Thanks @Olina-Zhang! We can definitely merge this then

**Review by lonitra — APPROVED** [review]
Thank you!

**Comment by MandiMan** [comment]
Tested this PR on .NET 9.0.0-preview.2.24076.2 with dlls built from winforms repo of main branch, no new issue found.

https://github.com/dotnet/winforms/assets/108860782/d1ca6c02-b396-42e9-b6ee-b2301dc9a4c3

**Pull request merged**